dataactcore/scripts/loadHistoricalFabs.py

```diff
@@ -177,6 +187,141 @@ def format_proper_casing(row, header):
+    if not row['agency_code']:
+        return None
+    awarding_sub_tier = sess.query(SubTierAgency). \
+        filter_by(sub_tier_agency_code=row['agency_code']).one()
+    return awarding_sub_tier.sub_tier_agency_name
+
+
+def derive_place_of_perform_county_na(row, sess):
+    if not row['principal_place_zip']:
+        ppop_code = row['principal_place_code'].upper()
+        ppop_state = None
+        if ppop_code[:2] == '00':
+            return None
+        elif re.match('^[0-9]{2}', ppop_code[:2]):
+            ppop_state = sess.query(States).filter_by(fips_code=ppop_code[:2]).one()
+        elif re.match('^[A-Z]{2}\*\*\d{3}$', ppop_code[:2]):
+            ppop_state = sess.query(States).filter_by(state_code=ppop_code[:2]).one()
+        if re.match('\*\*\d{3}$', ppop_code):
+            # getting county name
+            county_code = ppop_code[-3:]
+            county_info = sess.query(CountyCode).\
+                filter_by(county_number=county_code, state_code=ppop_state.state_code).first()
+            return county_info.county_name
+        elif re.match('\d{5}$', ppop_code) and not re.match('0{5}$', ppop_code):
+            # getting city and county name
+            city_code = ppop_code[-5:]
+            city_info = sess.query(CityCode).filter_by(city_code=city_code, state_code=ppop_state.state_code).first()
+            return city_info.county_name
+        return None
+    zip_five = row['principal_place_zip'][:5]
+    zip_four = None
+    if len(row['principal_place_zip']) > 5:
+        zip_four = row['principal_place_zip'][-4:]
+    if zip_four:
+        zip_info = sess.query(Zips). \
+            filter_by(zip5=zip_five, zip_last4=zip_four).first()
+    else:
+        zip_info = sess.query(Zips). \
+            filter_by(zip5=zip_five).first()
+    county_info = sess.query(CountyCode). \
+        filter_by(county_number=zip_info.county_number, state_code=zip_info.state_abbreviation).first()
+    return county_info.county_name
+
+
+def derive_place_of_performance_city(row, sess):
+    if not row['principal_place_zip']:
+        ppop_code = row['principal_place_code'].upper()
+        ppop_state = None
+        if ppop_code[:2] == '00':
+            return None
+        elif re.match('^[0-9]{2}', ppop_code[:2]):
+            ppop_state = sess.query(States).filter_by(fips_code=ppop_code[:2]).one()
+        elif re.match('^[A-Z]{2}\*\*\d{3}$', ppop_code[:2]):
+            ppop_state = sess.query(States).filter_by(state_code=ppop_code[:2]).one()
+        if re.match('\d{5}$', ppop_code) and not re.match('0{5}$', ppop_code):
+            # getting city and county name
+            city_code = ppop_code[-5:]
+            city_info = sess.query(CityCode).filter_by(city_code=city_code, state_code=ppop_state.state_code).first()
+            return city_info.feature_name
+        return None
+    zip_five = row['principal_place_zip'][:5]
+    city_info = sess.query(ZipCity).filter_by(zip_code=zip_five).one()
```

Comment:
Are we assuming this data and the zip data are in our DB? What if they didn't do a validation or something has changed since they validated and the zip or county isn't the same?